src/Text/Pandoc/Writers/XWiki.hs

```diff
@@ -0,0 +1,256 @@
+import Control.Monad.Reader (ReaderT, asks, local, runReaderT)
+import Data.Monoid
+import qualified Data.Set as Set
+import qualified Data.Text as Text
+import Data.Text (Text, concat, init, intercalate, pack, replace, split, unlines, unwords)
+import Text.Pandoc.Class (PandocMonad, report)
+import Text.Pandoc.Definition
+import Text.Pandoc.Logging
+import Text.Pandoc.Options
+import Text.Pandoc.Shared (escapeURI, isURI, linesToPara)
+import Text.Pandoc.Writers.MediaWiki (highlightingLangs)
+
+data WriterState = WriterState {
+  listLevel :: Text -- String at the beginning of items
+}
+
+type XWikiReader m = ReaderT WriterState m
+
+-- | Convert Pandoc to XWiki.
+writeXWiki :: PandocMonad m => WriterOptions -> Pandoc -> m Text
+writeXWiki _ (Pandoc _ blocks) =
+  let env = WriterState { listLevel = "" }
+  in do
+  body <- runReaderT (blockListToXWiki blocks) env
+  return $ body
+
+-- | Concatenates strings with line breaks between them.
+vcat :: [Text] -> Text
+vcat = intercalate "\n"
+
+blockListToXWiki :: PandocMonad m => [Block] -> XWikiReader m Text
+blockListToXWiki blocks =
+  fmap vcat $ mapM blockToXWiki blocks
+
+blockToXWiki :: PandocMonad m => Block -> XWikiReader m Text
+
+blockToXWiki Null = return ""
+
+-- TODO: handle this
+blockToXWiki (Div _ blocks) =
+  blockListToXWiki blocks
+
+blockToXWiki (Plain inlines) =
+  inlineListToXWiki inlines
+
+blockToXWiki (Para inlines) = do
+  contents <- inlineListToXWiki inlines
+  return $ contents <> "\n\n"
+
+blockToXWiki (LineBlock lns) =
+  blockToXWiki $ linesToPara lns
+
+blockToXWiki b@(RawBlock f str)
+  | f == Format "xwiki" = return $ pack str
+  | otherwise           = "" <$ report (BlockNotRendered b)
+
+blockToXWiki HorizontalRule = return "\n----\n"
+
+blockToXWiki (Header level _ inlines) = do
+  contents <- inlineListToXWiki inlines
+  let eqs = Text.replicate level "="
+  return $ eqs <> " " <> contents <> " " <> eqs <> "\n"
+
+-- XWiki doesn't appear to differentiate between inline and block-form code, so we delegate
+-- We do amend the text to ensure that the code markers are on their own lines, since this is a block
+blockToXWiki (CodeBlock attrs str) = do
+  contents <- inlineToXWiki (Code attrs ("\n" <> str <> "\n"))
+  return $ "\n" <> contents <> "\n"
+
+-- TODO: Figure out how to handle this better
+blockToXWiki (BlockQuote blocks) = do
+  blockText <- blockListToXWiki blocks
+  let quoteLines = split (== '\n') blockText
+  let prefixed = map ("> " <>) quoteLines
+  return $ vcat prefixed
+
+blockToXWiki (BulletList contents) = blockToXWikiList "*" $ contents
+
+blockToXWiki (OrderedList _ contents) = blockToXWikiList "1" $ contents
+
+blockToXWiki (DefinitionList items) = do
+  lev <- asks listLevel
+  contents <- local (\s -> s { listLevel = listLevel s <> ";" }) $ mapM definitionListItemToMediaWiki items
+  return $ vcat contents <> if Text.null lev then "\n" else ""
+
+-- TODO: support more features
+blockToXWiki (Table _ _ _ headers rows') = do
```

Comment:
Will this work even for complex table cells containing multiple blocks?
Might want to fall back to raw HTML for complex tables.
See other wiki writers.

Reply:
XWiki doesn't support raw html, unfortunately. I found this out while trying to fix RawBlock (below). I can take a look at how DokuWiki does non-raw table handling and see if I can adapt it.